knime_extension/src/nodes/locationanalysis.py

```diff
@@ -633,3 +681,548 @@ def getIndex(s):
+    The second (bottom) input should only include the demand ID column which is identical to the one from the top input, 
+    along with columns representing the distance matrix to candidate facilities (n).It has a dimension of m x (n+1).
+
+    The P-Median problem will be solved using the PuLP package.
+    """,
+    references={
+        "Pulp.Solver": "https://coin-or.github.io/pulp/guides/how_to_configure_solvers.html",
+    },
+)
+class PmedianSolverNode:
+    id_left = knext.ColumnParameter(
+        "Demand ID column from demand table",
+        "The column indicating the demand location IDs.",
+        # port_index=0,
+        column_filter=knut.is_int_or_string,
+        include_row_key=False,
+        include_none_column=False,
+    )
+    populationid = knext.ColumnParameter(
+        "Demand size column",
+        "The column for demand size (e.g., population) ",
+        # port_index=0,
+        column_filter=knut.is_numeric,
+        include_row_key=False,
+        include_none_column=False,
+    )
+
+    requiredid = knext.ColumnParameter(
+        "Distance column for required facilities",
+        "The travel cost to the nearest required facilities ",
+        # port_index=0,
+        column_filter=knut.is_numeric,
+        include_row_key=False,
+        include_none_column=True,
+    )
+
+    id_right = knext.ColumnParameter(
+        "Demand ID column from distance matrix table",
+        "The column indicating the demand location IDs.",
+        port_index=1,
+        column_filter=knut.is_int_or_string,
+        include_row_key=False,
+        include_none_column=False,
+    )
+    pnumber = knext.IntParameter(
+        "Optimum p number ",
+        "The optimum p number of facilities",
+        4,
+    )
+
+    def configure(self, configure_context, input_schema_1, input_schema_2):
+        return knext.Schema.from_columns(
+            [
+                knext.Column(knext.string(), _FACILITYID),
+                knext.Column(knext.int32(), _Chosen),
+            ]
+        )
+
+    def execute(self, exec_context: knext.ExecutionContext, input_1, input_2):
+        import pulp
+        import pandas as pd
+
+        df1, df2 = locationadata.load(input_1, self.id_left, input_2, self.id_right)
+        popu = df1[self.populationid].values.tolist()
+
+        cost_matrix, demand, candidate, p, nrequire = locationadata.process(
+            df1, df2, self.pnumber, self.requiredid
+        )
+
+        # Define the optimization problem
+        prob = pulp.LpProblem("p-media", pulp.LpMinimize)
+
+        # Define the decision variables
+        x, y = locationadata.ppulpxy(demand, candidate)
+
+        # Define the objective function
+        prob += pulp.lpSum(
+            [
+                x[(i, j)] * cost_matrix.iloc[i][j] * popu[i]
+                for i in range(demand)
+                for j in range(candidate)
+            ]
+        )
+
+        # Define the constraints
+        for i in range(demand):
```

Comment:
Add progress and cancellation

Reply:
Here is only for build xij list, can not be cancelled